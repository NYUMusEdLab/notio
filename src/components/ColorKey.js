import React, { Component } from "react";
import PropTypes from "prop-types";
import Star from "../assets/img/Star";
import MusicalStaff from "./MusicalStaff";

class ColorKey extends Component {
    constructor(props) {
        super(props);
        this.keyRef = React.createRef();
    }
    touchDown = e => {
        if (e.cancelable) {
            e.preventDefault();
        }
        if (this.props.isOn) {
            this.playNote(this.props.note);
        }
    };
    touchUp = e => {
        if (e.cancelable) {
            e.preventDefault(); // prevent default calling of mouse event after touch event
        }
        if (this.props.isOn) {
            this.releaseNote(this.props.note);
        }
    };

    clickedMouse = e => {
        if (this.props.isOn) {
            this.playNote(this.props.note);
        }
    };
    unClickedMouse = e => {
        if (this.props.isOn) {
            this.releaseNote(this.props.note);
        }
    };

    mouseEnter = e => {
        if (this.props.isOn && this.props.isMouseDown === true) {
            this.playNote(this.props.note);
        }
    };

    mouseLeave = e => {
        if (this.props.isOn && this.props.isMouseDown === true) {
            this.releaseNote(this.props.note);
        }
    };

    playNote = note => {
        this.props.noteOn(note);
    };

    releaseNote = note => {
        this.props.noteOff(note);
    };

    updateDimensions = () => {
        const myWidth = this.keyRef.current.clientWidth;
        this.setState({ myWidth: myWidth });
    };

    componentDidUpdate(prevProps) {
        if (
            this.props.showOffNotes !== prevProps.showOffNotes ||
            this.props.noteName !== prevProps.noteName
        ) {
            this.updateDimensions();
        }
    }

    componentDidMount() {
        this.keyRef.current.addEventListener("mouseenter", this.mouseEnter);
        this.keyRef.current.addEventListener("mouseleave", this.mouseLeave);
        //run it the first time
        this.updateDimensions();
        // and then run it on resize
        window.addEventListener("resize", this.updateDimensions);
    }

    componentWillUnmount() {
        // Make sure to remove the DOM listener when the component is unmounted.
        this.keyRef.current.removeEventListener("mouseenter", this.mouseEnter);
        this.keyRef.current.removeEventListener("mouseleave", this.mouseLeave);
        window.removeEventListener("resize", this.updateDimensions);
    }

    static getDerivedStateFromProps(nextProps, prevState) {
        return {
            isMouseDown: nextProps.isMouseDown
        };
    }

    render() {
        const {
            color,
            //offColor,
            keyColor,
            isOn,
            noteName,
            theme,
            trebleStaffOn,
            keyIndex
        } = this.props;

        
        let offKeyColorWithTheme;

        switch (theme) {
            case "light":
                offKeyColorWithTheme = '#ccc'; //offColor;
                break;
            case "dark":
                offKeyColorWithTheme = "#000";
                break;
            default:
                offKeyColorWithTheme = '#ccc'; //offColor;
        }

        const noteNames = noteName
            ? noteName.map(function (item, i) {
                item = item.toString();
                if (item.indexOf("b") === 2) {
                    //doubleflat
                    item = item.replace("bb", "\u1D12B");
                } else if (item.indexOf("b") === 1) {
                    //flat
                    item = item.replace("b", "\u266D");
                }
                return (
                    <div className="noteName" key={i}>
                        {item}
                    </div>
                );
            })
            : null;

        return (
            <div
                ref={this.keyRef}
                className={`color-key ${
                    this.state.clicked && isOn ? "active" : ""
                    } ${isOn ? "on" : "off"} ${
                    (this.props.note.includes("C") &&
                        !this.props.note.includes("#") &&
                        !this.props.note.includes("b")) ||
                        this.props.note.includes("B#")
                        ? ''/*"c-mark"*/
                        : ""
                    }
                    `}
                    style={{
                        height: this.props.pianoOn ? '70%' : '100%',
                        backgroundColor: isOn ? color : offKeyColorWithTheme,
                        opacity: theme === "dark" ? "0.7" : null
                    }}
                onMouseUp={this.unClickedMouse}
                onMouseDown={this.clickedMouse}
                onTouchStart={this.touchDown}
                onTouchEnd={this.touchUp}
                onMouseEnter={this.mouseEnter}
                onMouseLeave={this.mouseLeave}
            >
                <div
                    className={`note ${isOn ? "on" : "off"}`}
                    style={{
                        backgroundColor: this.state.clicked ? color : null,
                        
                    }}
                >
                    {noteNames}

                    {this.props.isMajorSeventh ? (
                        <div className="seventh">
                            <Star style={{ height: 50 }} />
                        </div>
                    ) : null}
                </div>
                {
                    trebleStaffOn && this.state.myWidth ? (
                        <MusicalStaff
                            width={this.state.myWidth}
                            note={this.props.note}
                            showOffNotes={this.props.showOffNotes}
                            keyIndex={keyIndex}
                            isOn={isOn}
                        />
                    ) : null
                }
            </div>
        )
    }
}


ColorKey.propTypes = {
    note: PropTypes.string,
    notation: PropTypes.array,
    noteName: PropTypes.array,
    color: PropTypes.string,
    //offcolor: PropTypes.string,
    keyColor: PropTypes.string,
    isOn: PropTypes.bool,
    root: PropTypes.string,
    isMajorSeventh: PropTypes.bool,
    keyIndex: PropTypes.number
};

export default ColorKey;