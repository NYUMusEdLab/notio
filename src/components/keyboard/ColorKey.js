import React, { Component } from "react";
import PropTypes from "prop-types";
import Star from "../../assets/img/Star";
import MusicalStaff from "../musicScore/MusicalStaff";
import Color from "color";

class ColorKey extends Component {
  constructor(props) {
    super(props);
    this.keyRef = React.createRef();
    this.initColor(props.color);
    this.state = {
      isMouseDown: false,
      _color: this._colorInit,
    };
  }
  touchDown = (e) => {
    if (e.cancelable) {
      e.preventDefault();
    }
    if (this.props.toneIsInScale) {
      this.playNote(this.props.note);
    }
  };
  touchUp = (e) => {
    if (e.cancelable) {
      e.preventDefault(); // prevent default calling of mouse event after touch event
    }
    if (this.props.toneIsInScale) {
      this.releaseNote(this.props.note);
    }
  };

  //TODO: check if this is used or can be used to fix the unresponsive keys color when pressed
  onMouseOver = (e) => {
    if (this.props.toneIsInScale) {
      // this.setState((state) => {
      //   return { _color: this._colorActive, }
      // })
    }
  };

  //TODO: check if this is used or can be used to fix the unresponsive keys color when pressed
  onMouseOut = (e) => {
    if (this.props.toneIsInScale) {
      // this.setState((state) => {
      //   return { _color: this._colorInit, }
      // })
    }
  };

  clickedMouse = (e) => {
    if (this.props.toneIsInScale) {
      this.setState((state) => {
        return {
          _color: "linear-gradient(180deg, rgba(255,255,255,0) 20%,  100% 100%)",
        };
      });
      this.playNote(this.props.note);
    }
  };
  unClickedMouse = (e) => {
    if (this.props.toneIsInScale) {
      this.setState((state) => {
        return { _color: this._colorInit };
      });
      this.releaseNote(this.props.note);
    }
  };

  mouseEnter = (e) => {
    if (this.props.toneIsInScale && this.props.isMouseDown === true) {
      this.playNote(this.props.note);
    }
  };

  mouseLeave = (e) => {
    if (this.props.toneIsInScale && this.props.isMouseDown === true) {
      this.releaseNote(this.props.note);
    }
  };

  playNote = (note) => {
    this.props.noteOnHandler(note);
  };

  releaseNote = (note) => {
    this.props.noteOffHandler(note);
  };

  updateDimensions = () => {
    const myWidth = this.keyRef.current.clientWidth;
    this.setState({ myWidth: myWidth });
  };

  initColor = (color) => {
    this._colorInit = Color(color);
  };

  componentDidUpdate(prevProps) {
    if (
      this.props.showOffNotes !== prevProps.showOffNotes ||
      this.props.noteName !== prevProps.noteName
    ) {
      this.updateDimensions();
    }

    if (this.props.color !== prevProps.color) {
      this.initColor(this.props.color);
      this.setState({ _color: this._colorInit });
    }
  }

  componentDidMount() {
    this.keyRef.current.addEventListener("mouseenter", this.mouseEnter);
    this.keyRef.current.addEventListener("mouseleave", this.mouseLeave);
    //run it the first time
    this.updateDimensions();
    // and then run it on resize
    window.addEventListener("resize", this.updateDimensions);
  }

  componentWillUnmount() {
    // Make sure to remove the DOM listener when the component is unmounted.
    this.keyRef.current.removeEventListener("mouseenter", this.mouseEnter);
    this.keyRef.current.removeEventListener("mouseleave", this.mouseLeave);
    window.removeEventListener("resize", this.updateDimensions);
  }

  static getDerivedStateFromProps(nextProps, prevState) {
    return {
      isMouseDown: nextProps.isMouseDown,
    };
  }

  render() {
    const {
      color,
      toneIsInScale,
      pianoOn,
      noteName,
      theme,
      trebleStaffOn,
      keyIndex,
      note,
      showOffNotes,
      isMajorSeventh,
      extendedKeyboard,
      clef,
    } = this.props;

    // console.log("ColorKey note", note);

    const noteNames = noteName
      ? noteName.map(function (item, i) {
          var name = item.value.toString();
          const regexFlat = /[b]/gi;
          const regexSharp = /[#]/gi;

          let flats = name.match(regexFlat);
          let sharps = name.match(regexSharp);
          if (flats && flats.length === 2) {
            //doubleflat
            name = name.replace("bb", "ùÑ´");
            //item = item.replace("bb", "\u1D12B");
          } else if (flats && flats.length === 1) {
            //flat
            name = name.replace("b", "\u266D");
          } else if (sharps && sharps.length === 2) {
            //doubleSharp
            name = name.replace("##", "√ó"); //ùÑ™
          }

          return (
            <div className={`noteName noteWrapper--${item.key}`} key={i}>
              {name}
            </div>
          );
        })
      : null;

    return (
      <div
        ref={this.keyRef}
        data-testid={"ColorKey:" + note}
        className={`color-key`}
        style={{
          height: pianoOn ? "70%" : "100%",
          //background is set to color if tone is in scale or the gradient app background if it is not in scale
          // @ts-ignore
          background: toneIsInScale
            ? this.state._color
            : "linear-gradient(356deg, rgba(249,247,223,1) 0%, rgba(235,227,245,1) 100%)",
          opacity: theme === "dark" ? "0.7" : null,
        }}
        onMouseUp={this.unClickedMouse}
        onMouseDown={this.clickedMouse}
        onMouseOver={this.onMouseOver}
        onMouseOut={this.onMouseOut}
        onTouchStart={this.touchDown}
        onTouchEnd={this.touchUp}
        onMouseEnter={this.mouseEnter}
        onMouseLeave={this.mouseLeave}>
        <div
          className={`noteWrapper note ${toneIsInScale ? "on" : "off"}`}
          style={{
            marginBottom: pianoOn ? "0%" : "16vh", //TODO: find better solution, this will not work on all screen sizes
            backgroundColor: this.state.clicked ? color : null,
            top: extendedKeyboard ? "17%" : "10%",
          }}>
          {noteNames}

          {isMajorSeventh ? (
            <div className="seventh">
              <Star style={{ height: 50 }} />
            </div>
          ) : null}
        </div>
        {trebleStaffOn && this.state.myWidth ? (
          <MusicalStaff
            width={this.state.myWidth}
            note={note}
            showOffNotes={showOffNotes}
            keyIndex={keyIndex}
            toneIsInScale={toneIsInScale}
            extendedKeyboard={extendedKeyboard}
            clef={clef}
          />
        ) : null}
      </div>
    );
  }
}

ColorKey.propTypes = {
  note: PropTypes.string,
  notation: PropTypes.array,
  noteName: PropTypes.array,
  color: PropTypes.string,
  keyColor: PropTypes.string,
  toneIsInScale: PropTypes.bool,
  root: PropTypes.string,
  isMajorSeventh: PropTypes.bool,
  keyIndex: PropTypes.number,
  extendedKeyboard: PropTypes.bool,
};

export default ColorKey;
