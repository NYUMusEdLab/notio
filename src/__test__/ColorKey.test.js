import React from "react";
import { render, fireEvent, screen } from "@testing-library/react";
import ColorKey from "../components/keyboard/ColorKey";

const testProps = {
  clef: "treble",
  color: "#ff0000",
  extendedKeyboard: false,
  index: 0,
  isActive: false,
  isMouseDown: false,
  keyColor: "white",
  keyIndex: undefined,
  note: "C4",
  noteName: [],
  noteNameEnglish: "C",
  noteOffHandler: (note) => {},
  noteOnHandler: (note) => {},
  pianoOn: true,
  root: "C",
  synth: {},
  theme: "light",
  toneIsInScale: true,
  trebleStaffOn: true,
};

describe("ColorKey", () => {
  it("should render correctly", () => {
    const { container } = render(<ColorKey {...testProps} />);
    expect(container).toMatchSnapshot();
  });

  it("should call touchUp when key is touched", () => {
    const touchUpMock = jest.fn();
    render(<ColorKey {...testProps} touchUp={touchUpMock} />);

    const colorKey = screen.getByTestId("ColorKey:C4");
    fireEvent.touchStart(colorKey);
    fireEvent.touchEnd(colorKey);

    expect(touchUpMock).toHaveBeenCalled();
  });

  it("should call onMouseOver when key is hovered", () => {
    const onMouseOverMock = jest.fn();
    render(<ColorKey {...testProps} onMouseOver={onMouseOverMock} />);

    const colorKey = screen.getByTestId("ColorKey:C");
    fireEvent.mouseOver(colorKey);

    expect(onMouseOverMock).toHaveBeenCalled();
  });

  it("should call clickMouse when key is clicked", () => {
    const clickMouseMock = jest.fn();
    render(<ColorKey {...testProps} clickMouse={clickMouseMock} />);

    const colorKey = screen.getByTestId("ColorKey:C");
    fireEvent.click(colorKey);

    expect(clickMouseMock).toHaveBeenCalled();
  });

  it("should call componentDidMount", () => {
    const spy = jest.spyOn(ColorKey.prototype, "componentDidMount");
    render(<ColorKey {...testProps} />);

    expect(spy).toHaveBeenCalled();
  });

  it("should call componentDidUpdate", () => {
    const spy = jest.spyOn(ColorKey.prototype, "componentDidUpdate");
    render(<ColorKey {...testProps} />);
    expect(spy).toHaveBeenCalled();
  });
});

// import React from "react";
// import { render, fireEvent, screen } from "@testing-library/react";
// import "@testing-library/jest-dom/extend-expect";
// import ColorKey from "./ColorKey";

// describe("ColorKey", () => {
//   it("should render the correct note name", () => {
//     const noteName = "C";
//     render(<ColorKey noteName={noteName} />);
//     const colorKey = screen.getByTestId(`ColorKey:${noteName}`);
//     expect(colorKey).toHaveTextContent(noteName);
//   });

//   it("should change color on touchUp", () => {
//     const noteName = "C";
//     render(<ColorKey noteName={noteName} />);
//     const colorKey = screen.getByTestId(`ColorKey:${noteName}`);
//     fireEvent.touchStart(colorKey);
//     expect(colorKey).toHaveStyle("background-color: var(--color-key-active)");
//     fireEvent.touchEnd(colorKey);
//     expect(colorKey).toHaveStyle("background-color: var(--color-key-inactive)");
//   });

//   it("should change color on mouseOver and mouseOut", () => {
//     const noteName = "C";
//     render(<ColorKey noteName={noteName} />);
//     const colorKey = screen.getByTestId(`ColorKey:${noteName}`);
//     fireEvent.mouseOver(colorKey);
//     expect(colorKey).toHaveStyle("background-color: var(--color-key-active)");
//     fireEvent.mouseOut(colorKey);
//     expect(colorKey).toHaveStyle("background-color: var(--color-key-inactive)");
//   });

//   it("should change color on click", () => {
//     const noteName = "C";
//     render(<ColorKey noteName={noteName} />);
//     const colorKey = screen.getByTestId(`ColorKey:${noteName}`);
//     fireEvent.click(colorKey);
//     expect(colorKey).toHaveStyle("background-color: var(--color-key-active)");
//     fireEvent.click(colorKey);
//     expect(colorKey).toHaveStyle("background-color: var(--color-key-inactive)");
//   });

//   it("should call componentDidMount lifecycle method", () => {
//     const noteName = "C";
//     const spy = jest.spyOn(ColorKey.prototype, "componentDidMount");
//     render(<ColorKey noteName={noteName} />);
//     expect(spy).toHaveBeenCalled();
//   });
//   it("should call componentWillUnmount lifecycle method", () => {
//     const noteName = "C";
//     const spy = jest.spyOn(ColorKey.prototype, "componentWillUnmount");
//     const { unmount } = render(<ColorKey note={noteName} />);
//     unmount();
//     expect(spy).toHaveBeenCalled();
//   });
// });
