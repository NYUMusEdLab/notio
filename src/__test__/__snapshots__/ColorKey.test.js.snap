// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ColorKey should render correctly 1`] = `
<div>
  <div
    class="color-key on"
    data-testid="ColorKey:C4"
    style="height: 70%; background: rgb(255, 0, 0);"
  >
    <div
      class="noteWrapper note on"
      style="margin-bottom: 0%; top: 10%;"
    />
  </div>
</div>
`;
